### PR TITLE
Flag non-nullable functions in `if` statements as errors (tree walk version)

### PR DESCRIPTION
**Pull request by jwbay**
This is an alternate version of #32802 -- see that one for context.

This version attempts to prevent false positives by doing a syntax walk to see if the checked function is used, as opposed to relying on a heuristic where the function returns a boolean. The first commit is the same for both PR; the second adds filtering that differs.

Opened as a separate PR by request (cc @RyanCavanaugh)

Notably, this change flagged a couple things in the compiler itself. Commented below for both.

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -1362,18 +1362,16 @@ namespace ts {
             // try to verify results of module resolution
             for (const { oldFile: oldSourceFile, newFile: newSourceFile } of modifiedSourceFiles) {
                 const newSourceFilePath = getNormalizedAbsolutePath(newSourceFile.originalFileName, currentDirectory);
-                if (resolveModuleNamesWorker) {
```
Comment (jwbay) [inline src/compiler/program.ts]: Looks like `resolveModuleNamesWorker` is always assigned above, so this was a dead check

File: src/compiler/core.ts [hunk 2]
```diff
@@ -1653,7 +1653,7 @@ namespace ts {
      */
     export function compose<T>(...args: ((t: T) => T)[]): (t: T) => T;
     export function compose<T>(a: (t: T) => T, b: (t: T) => T, c: (t: T) => T, d: (t: T) => T, e: (t: T) => T): (t: T) => T {
-        if (e) {
+        if (!!e) {
```
Comment (jwbay) [inline src/compiler/core.ts]: This is a pretty good example of what TS would now consider suspicious-enough code to error on. Making `e` optional would also have fixed this.

**Comment by orta** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt

**Comment by RyanCavanaugh** [comment]
The RWC diff is unrelated.

The User Test Suite diff is a false positive, but not a bad one: https://github.com/npm/cli/blob/latest/test/tap/check-permissions.js#L27

**Comment by RyanCavanaugh** [comment]
@typescript-bot perf test this

**Comment by orta** [comment]
Hrm, none of these builds have accessible logs. 

I bet it's likely that community projects will need changes like:

```diff
-        if (e) {
+       if (!!e) {
```

Will re-run and take a look.

@typescript-bot test this
@typescript-bot user test this

**Comment by orta** [comment]
OK @jwbay - to get this mergeable you'll need to accept https://github.com/jwbay/TypeScript/pull/3

**Comment by orta** [comment]
OK, perf looks acceptable to me.

The idea is a good one, and I feel good about the implementation - thanks @jwbay!

**Comment by user753** [comment]
@jwbay 
Is there any reason why it affects only functions and not arrays, records, etc?

```
function foo(array: string[]) {
  if (array) {
    return true
  }
}

function bar(record: Record<string, string>) {
  if (record) {
    return true
  }
}
```

**Comment by RyanCavanaugh** [comment]
@user753 a few major problems appeared when we tried that approach.

Array access is optimistically assumed to be in-bounds, so code like this gets incorrectly flagged as an error. The same thing happens for `Map.get` and a few other APIs that are assumed to "always return non-undefined values when used correctly".
```ts
const arr = [rec, rec, rec];
const el = arr[someIndex];
if (el) {
  // do something
}
```

Separately, many definition files were written prior to the introduction of `strictNullChecks`, so correct `null`/`undefined` checks on properties from those definitions would get incorrectly flagged as an error.

Ultimately the very narrow check implemented here was the only one that didn't generate a huge amount of false positives.

**Comment by user753** [comment]
@RyanCavanaugh Thanks. 
```
const arr = [func, func, func];
const el = arr[someIndex];
if (el) {
  // do something
}
```
Do I understand it correctly that It isn't a problem for functions because a function inside an array isn't a common use case?

**Comment by RyanCavanaugh** [comment]
Normally if that does happen, a call to `el()` appears in the body of the `if`, which is the additional check that this PR implements. If `el` *isn't` called, that'll be a false positive.

**Comment by RyanCavanaugh** [comment]
@jwbay this might be my favorite check this release. https://github.com/microsoft/vscode/blob/ecba37c6ef3aab85b7a9b15afb3610bbe4eaac8c/extensions/css-language-features/server/src/cssServerMain.ts#L78

**Comment by travigd** [comment]
It would be great if this was extended to functions that return promises as well.

In the announcement for 3.7 (which I'm **thrilled** about!!!), you include this code example.
```ts
function doAdminThing(user: User) {
    if (user.isAdministrator) {
    //  ~~~~~~~~~~~~~~~~~~~~
    // error! This condition will always return true since the function is always defined.
    //        Did you mean to call it instead?t
```

It would be incredible if this was also extended to raise an error when the function returns a promise and it's not awaited before using it in a condition (I've personally been bitten by this exact bug before).
```ts
class User {
    // ...
    async isAdministrator() { /* ... */ }
}
function doAdminThing(user: User) {
    if (user.isAdministrator()) {
    //  ~~~~~~~~~~~~~~~~~~~~~~
    // error! This condition will always return true since a promise is always truthy.
    //        Did you mean to await it instead?
```

I tried searching for issues related to this but couldn't find any (it's possible that I just missed it because there are so many issues and I'm not sure what keywords to search).

I've definitely been bitten by a bug where I did something like this:
```ts
async function passwordMatches(email: string, password: string) {
  /* return true if the password matches the result in the database */
}

if (passwordMatches("foo@bar.gov", "passw0rd")) {
  /* issue auth token */
}
```

**Comment by omidkrad** [comment]
Regarding Uncalled Function Checks, I think it'd be pretty safe to also error on this:

```typescript
function process() {
  console.log('Processing...');
}

process;
```

**Comment by nebrelbug** [comment]
BTW this is breaking all of my polyfills, like checks for `if (String.prototype.trimLeft) {}`. Any way to disable this?

**Comment by vinothbabu** [comment]
The problem with this is, it's not allowing any of the valid ways to pass. As @nebrelbug pointed above. 

`if (String.prototype.trimLeft) {} //fails`
`if (typeof String.prototype.trimLeft === 'function') {} //fails`
`if ('trimLeft' in String.prototype) //fails `

We had to adapt to this, which passes

`// eslint-disable-next-line no-extra-boolean-cast`
`if (!!String.prototype.trimLeft) `

**Pull request merged**